### PR TITLE
infoschema, util: refresh statement summary table periodically

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve? 
Refresh the table `events_statements_summary_by_digest` periodically.

### What is changed and how it works?
If the current summary exists for over `tidb_stmt_summary_refresh_interval` seconds, then replace it with a new one.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```
mysql> select * from events_statements_summary_by_digest limit 1\G
*************************** 1. row ***************************
       SUMMARY_BEGIN_TIME: 2019-11-22 13:00:00
                STMT_TYPE: select
              SCHEMA_NAME: performance_schema
                   DIGEST: 688edff0859079e96540c6161b08dde9ab5f8654bd2755a0dbca27de5e06f93e
              DIGEST_TEXT: select * from events_statements_summary_by_digest
              TABLE_NAMES: performance_schema.events_statements_summary_by_digest
              INDEX_NAMES: NULL
              SAMPLE_USER: root@127.0.0.1
               EXEC_COUNT: 1
              SUM_LATENCY: 749473
              MAX_LATENCY: 749473
              MIN_LATENCY: 749473
              AVG_LATENCY: 749473
        AVG_PARSE_LATENCY: 56405
        MAX_PARSE_LATENCY: 56405
      AVG_COMPILE_LATENCY: 444081
      MAX_COMPILE_LATENCY: 444081
             COP_TASK_NUM: 0
     AVG_COP_PROCESS_TIME: 0
     MAX_COP_PROCESS_TIME: 0
  MAX_COP_PROCESS_ADDRESS: NULL
        AVG_COP_WAIT_TIME: 0
        MAX_COP_WAIT_TIME: 0
     MAX_COP_WAIT_ADDRESS: NULL
         AVG_PROCESS_TIME: 0
         MAX_PROCESS_TIME: 0
            AVG_WAIT_TIME: 0
            MAX_WAIT_TIME: 0
         AVG_BACKOFF_TIME: 0
         MAX_BACKOFF_TIME: 0
           AVG_TOTAL_KEYS: 0
           MAX_TOTAL_KEYS: 0
       AVG_PROCESSED_KEYS: 0
       MAX_PROCESSED_KEYS: 0
        AVG_PREWRITE_TIME: 0
        MAX_PREWRITE_TIME: 0
          AVG_COMMIT_TIME: 0
          MAX_COMMIT_TIME: 0
   AVG_GET_COMMIT_TS_TIME: 0
   MAX_GET_COMMIT_TS_TIME: 0
  AVG_COMMIT_BACKOFF_TIME: 0
  MAX_COMMIT_BACKOFF_TIME: 0
    AVG_RESOLVE_LOCK_TIME: 0
    MAX_RESOLVE_LOCK_TIME: 0
AVG_LOCAL_LATCH_WAIT_TIME: 0
MAX_LOCAL_LATCH_WAIT_TIME: 0
           AVG_WRITE_KEYS: 0
           MAX_WRITE_KEYS: 0
           AVG_WRITE_SIZE: 0
           MAX_WRITE_SIZE: 0
     AVG_PREWRITE_REGIONS: 0
     MAX_PREWRITE_REGIONS: 0
            AVG_TXN_RETRY: 0
            MAX_TXN_RETRY: 0
            BACKOFF_TYPES: NULL
                  AVG_MEM: 0
                  MAX_MEM: 0
        AVG_AFFECTED_ROWS: 0
               FIRST_SEEN: 2019-11-22 13:22:53
                LAST_SEEN: 2019-11-22 13:22:53
        QUERY_SAMPLE_TEXT: select * from events_statements_summary_by_digest
```

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release note

 - Refresh the system table `events_statements_summary_by_digest` periodically.

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -418,6 +421,7 @@ const (
 	DefWaitSplitRegionTimeout          = 300 // 300s
 	DefTiDBEnableNoopFuncs             = false
 	DefTiDBAllowRemoveAutoInc          = false
+	DefTiDBStmtSummaryRefreshInterval  = 1800
```
Comment (zimulala) [inline sessionctx/variable/tidb_vars.go]:
```suggestion
	DefTiDBStmtSummaryRefreshInterval  = 1800 // 1800s
```

File: util/stmtsummary/statement_summary.go [hunk 2]
```diff
@@ -58,14 +59,23 @@ type stmtSummaryByDigestMap struct {
...
+	// sysVars encapsulates system variables needed to control statement summary.
+	sysVars struct {
 		sync.RWMutex
 		// enabled indicates whether statement summary is enabled in current server.
 		sessionEnabled string
 		// setInSession indicates whether statement summary has been set in any session.
 		globalEnabled string
+		// XXXRefreshInterval indicates the refresh interval of summaries.
+		// It must be > 0.
+		sessionRefreshInterval string
```
Comment (AilinKid) [inline util/stmtsummary/statement_summary.go]: If we want > 0, why not use unsigned int?

File: util/stmtsummary/statement_summary.go [hunk 3]
```diff
@@ -76,6 +86,8 @@ var StmtSummaryByDigestMap = newStmtSummaryByDigestMap()
 type stmtSummaryByDigest struct {
 	// It's rare to read concurrently, so RWMutex is not needed.
 	sync.Mutex
+	// Each summary is summarized between [beginTime, beginTime + interval]
```
Comment (AilinKid) [inline util/stmtsummary/statement_summary.go]:
```suggestion
	// Each summary is summarized between [beginTime, beginTime + interval].
```

File: util/stmtsummary/statement_summary.go (hunk 2 above)
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: If it's an empty string, then we'll ignore it and use globalRefreshInterval.
This kind of operations was discussed with zhou sir, setting session variable to an empty string indicates unsetting session variable and then use global variable.

File: sessionctx/variable/tidb_vars.go (hunk 1 above)
Comment (djshow832) [reply sessionctx/variable/tidb_vars.go]: Addressed.

File: sessionctx/variable/varsutil.go [hunk 4]
```diff
@@ -628,6 +628,18 @@ func ValidateSetSystemVar(vars *SessionVars, name string, value string) (string,
...
+			return "", nil
+		}
+		v, err := strconv.Atoi(value)
+		if err != nil {
+			return value, ErrWrongTypeForVar.GenWithStackByArgs(name)
+		}
+		if v <= 0 {
+			return value, ErrWrongValueForVar.GenWithStackByArgs(name, value)
+		}
+		return value, nil
```
[the remark this answers is not in the document]
Comment (djshow832) [reply sessionctx/variable/varsutil.go]: Addressed.

File: util/stmtsummary/statement_summary.go (hunk 3 above)
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed.

File: util/stmtsummary/statement_summary.go (hunk 2 above)
Comment (AilinKid) [reply util/stmtsummary/statement_summary.go]: I see.

File: util/stmtsummary/statement_summary.go [hunk 5]
```diff
@@ -488,11 +566,16 @@ func (ssbd *stmtSummaryByDigest) add(sei *StmtExecInfo) {
...
+func (ssbd *stmtSummaryByDigest) toDatum(oldestBeginTime int64) []types.Datum {
 	ssbd.Lock()
 	defer ssbd.Unlock()
 
+	if ssbd.beginTime < oldestBeginTime {
+		return nil
+	}
+
 	return types.MakeDatums(
+		types.Time{Time: types.FromGoTime(time.Unix(ssbd.beginTime, 0)), Type: mysql.TypeTimestamp},
```
Comment (AilinKid) [inline util/stmtsummary/statement_summary.go]: Will time zone take effects?

Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: No.

File: util/stmtsummary/statement_summary.go [hunk 6]
```diff
@@ -58,14 +59,23 @@ type stmtSummaryByDigestMap struct {
 	// It's rare to read concurrently, so RWMutex is not needed.
 	sync.Mutex
 	summaryMap *kvcache.SimpleLRUCache
+	// These fields are used for rolling summary.
+	beginTimeForCurInterval int64
+	lastCheckExpireTime     int64
 
-	// enabledWrapper encapsulates variables needed to judge whether statement summary is enabled.
-	enabledWrapper struct {
+	// sysVars encapsulates system variables needed to control statement summary.
```
[the remark this answers is not in the document]
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed.

File: util/stmtsummary/statement_summary.go [hunk 7]
```diff
@@ -58,14 +59,23 @@ type stmtSummaryByDigestMap struct {
...
+	sysVars struct {
 		sync.RWMutex
 		// enabled indicates whether statement summary is enabled in current server.
 		sessionEnabled string
 		// setInSession indicates whether statement summary has been set in any session.
 		globalEnabled string
+		// XXXRefreshInterval indicates the refresh interval of summaries.
```
[the remark this answers is not in the document]
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: `XXXRefreshInterval` indicates `sessionRefreshInterval` + `globalRefreshInterval` + `refreshInterval`. Is there any better way to express?

[the remark this answers is not in the document]
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed.

File: util/stmtsummary/statement_summary.go [hunk 8]
```diff
@@ -167,16 +179,21 @@ type StmtExecInfo struct {
...
-	ssMap.enabledWrapper.globalEnabled = ""
+	// sysVars.defaultEnabled will be initialized in package variable.
+	ssMap.sysVars.sessionEnabled = ""
+	ssMap.sysVars.globalEnabled = ""
 	return ssMap
 }
 
 // AddStatement adds a statement to StmtSummaryByDigestMap.
 func (ssMap *stmtSummaryByDigestMap) AddStatement(sei *StmtExecInfo) {
+	// All times are count in seconds.
```
[the remark this answers is not in the document]
Comment (djshow832) [reply util/stmtsummary/statement_summary.go]: Addressed.

**Pull request merged**